str0m/src/session.rs

```diff
@@ -59,6 +59,11 @@ pub(crate) struct Session {
 
     enable_twcc_feedback: bool,
 
+    /// A pacer for sending RTP at specific rate.
+    pacer: Option<LeakyBucketPacer>,
+    /// A SRTP packet that is being held because the pacer prevented it from being sent.
+    held_srtp_packet: Option<(Vec<u8>, u64)>,
```

Comment:
Both the libwebrtc pacer and the one I wrote work in terms of the pacer being a queue.  You send things into it and then things come out of it.  I think that might be a better way to model this.  I guess you could put the queue here, but you probably don't want a queue that's only 1 packet.  That will lead to a lot of dropped packets with a burst such as can happen with a big key frame or screenshare where this is suddenly a lot of motion.

Reply:
It seems like it would be easy to end up with the logic being spread around quite a bit.  At some point it's likely you'll want a tight coupling between knowing what is queued and deciding what to do about it (how quickly to pace out, what things to drop and not drop, how much to queue before taking action).

Reply:
In light of today's discussions, I see the reason to have one. I would image putting some generic queue here with operations for reordering/dropping etc. My main concern previously was to make it the same component as Pacer.

Oh, and also I want to poll one packet out at a time rather than polling Vec<Packet>.

Reply:
Looking at this more and further reading libWebRTC's code I do think we want to move the queues from the many `PacketizingBuffer`s into the pacer. The pacer needs to have knowledge of queue sizes as well as being able to pick out different media. For example, consider the 40ms burst interval I discovered above, this is not workable for audio, hence libWebRTC handles this by not pacing audio. I see they also have a field trial to not pace retransmissions, which makes sense for similar reasons.